lib/connection_pool.rb

```diff
@@ -53,22 +53,28 @@ def initialize(options = {}, &block)
   end
 
   def with(options = {})
+    success = false # hoisted
     conn = checkout(options)
     begin
-      yield conn
+      (yield conn).tap do
+        success = true # means the connection wasn't interrupted
+      end
     ensure
-      checkin
+      if success
+        # everything is roses, we can safely check the connection back in
+        checkin
+      else
+        @available.discard!(pop_connection)
```

Comment:
My guess is that you're throwing an exception while the connection is held. The problem here is really that `connection_pool` is one level too high to address the `Timeout::timeout` problem - `connection_pool` can't know if the block failed while the connection was actually being used or just before or after, so it errs on the side of eagerly discarding connections which might still be healthy:

```ruby
pool = ConnectionPool::Wrapper.new(:size => 1) { retrieve_connection }

poo.with do |conn|
  conn.do_stuff
  # this marks the connection as corrupted because
  # `with` cannot know if the error came during a
  # method on the connection or not
  raise 'throw away the connection'
end
```